src/libopensc/card-jpki.c

```diff
@@ -0,0 +1,344 @@
+	sc_file_add_acl_entry(mf, SC_AC_OP_LOCK, SC_AC_NEVER, 0);
+	sc_file_add_acl_entry(mf, SC_AC_OP_DELETE, SC_AC_NEVER, 0);
+	sc_file_add_acl_entry(mf, SC_AC_OP_CREATE, SC_AC_NEVER, 0);
+	drvdata->mf = mf;
+
+	card->name = "jpki";
+	card->drv_data = drvdata;
+
+	flags = SC_ALGORITHM_RSA_HASH_NONE | SC_ALGORITHM_RSA_PAD_PKCS1;
+	_sc_card_add_rsa_alg(card, 2048, flags, 0);
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int
+jpki_select_file(struct sc_card *card,
+				 const struct sc_path *path, struct sc_file **file_out)
+{
+	struct jpki_private_data *drvdata = DRVDATA(card);
+	int rc;
+	sc_apdu_t apdu;
+	struct sc_file *file = NULL;
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+	sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,
+			 "jpki_select_file: path=%s, len=%d\n",
+			 sc_print_path(path), path->len);
+	if (path->len == 2 && memcmp(path->value, "\x3F\x00", 2) == 0) {
+		drvdata->selected = SELECT_MF;
+		if (file_out) {
+			sc_file_dup(file_out, drvdata->mf);
+			if (*file_out == NULL) {
+				return SC_ERROR_OUT_OF_MEMORY;
+			}
+		}
+		return 0;
+	}
+
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0xA4, 0, 0);
+	switch (path->type) {
+	case SC_PATH_TYPE_FILE_ID:
+		apdu.p1 = 2;
+		break;
+	case SC_PATH_TYPE_DF_NAME:
+		apdu.p1 = 4;
+		break;
+	default:
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+	}
+	apdu.p2 = 0x0C;
+	apdu.data = path->value;
+	apdu.datalen = path->len;
+	apdu.lc = path->len;
+
+	rc = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(card->ctx, rc, "APDU transmit failed");
+	rc = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	LOG_TEST_RET(card->ctx, rc, "SW Check failed");
+	if (file_out) {
+		file = sc_file_new();
+		if (!file) {
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+		}
+		file->path = *path;
+		file->size = 2048;
+		*file_out = file;
+	}
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int
+jpki_read_binary(sc_card_t * card, unsigned int idx,
+				 u8 * buf, size_t count, unsigned long flags)
+{
+	struct sc_card_driver *iso_drv = sc_get_iso7816_driver();
+	const struct sc_card_operations *iso_ops = iso_drv->ops;
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	return iso_ops->read_binary(card, idx, buf, count, flags);
+}
+
+static int
+jpki_pin_cmd(sc_card_t * card, struct sc_pin_cmd_data *data, int *tries_left)
+{
+	int rc;
+	sc_path_t path;
+	sc_apdu_t apdu;
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	if (tries_left) {
+		*tries_left = -1;
+	}
+
+	switch (data->pin_reference) {
+	case 1:
+		sc_format_path("0018", &path);
+		path.type = SC_PATH_TYPE_FILE_ID;
+		rc = sc_select_file(card, &path, NULL);
+		break;
+	case 2:
+		sc_format_path("001b", &path);
+		path.type = SC_PATH_TYPE_FILE_ID;
+		rc = sc_select_file(card, &path, NULL);
+		break;
+	default:
+		sc_log(card->ctx, "Unknown PIN reference: %d", data->pin_reference);
+		return SC_ERROR_INVALID_ARGUMENTS;
+	}
+	LOG_TEST_RET(card->ctx, rc, "SELECT_FILE error");
+
+	switch (data->cmd) {
```

Comment:
Really? I bet this is possible, but it's OK if you better like to do your own implementation to keep things simple (though you will have to carry the burden of maintaining this stuff :wink: ).